src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/CustomAttributeExtensions.cs

```diff
@@ -0,0 +1,121 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection.Metadata;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
+
+namespace ILCompiler
+{
+    public static class CustomAttributeExtensions
+    {
+        public static IEnumerable<CustomAttributeValue<TypeDesc>> GetDecodedCustomAttributesForEntity(this TypeSystemEntity entity, string attributeNamespace, string attributeName)
+        {
+            switch (entity)
+            {
+                case MethodDesc method:
+                    var ecmaMethod = method.GetTypicalMethodDefinition() as EcmaMethod;
+                    if (ecmaMethod == null)
+                        return Enumerable.Empty<CustomAttributeValue<TypeDesc>>();
+                    return ecmaMethod.GetDecodedCustomAttributes(attributeNamespace, attributeName);
+                case MetadataType type:
+                    var ecmaType = type.GetTypeDefinition() as EcmaType;
+                    if (ecmaType == null)
+                        return Enumerable.Empty<CustomAttributeValue<TypeDesc>>();
+                    return ecmaType.GetDecodedCustomAttributes(attributeNamespace, attributeName);
+                case FieldDesc field:
+                    var ecmaField = field.GetTypicalFieldDefinition() as EcmaField;
+                    if (ecmaField == null)
+                        return Enumerable.Empty<CustomAttributeValue<TypeDesc>>();
+                    return ecmaField.GetDecodedCustomAttributes(attributeNamespace, attributeName);
+                case PropertyPseudoDesc property:
+                    return property.GetDecodedCustomAttributes(attributeNamespace, attributeName);
+                case EventPseudoDesc @event:
+                    return @event.GetDecodedCustomAttributes(attributeNamespace, attributeName);
+                case ModuleDesc module:
+                    if (module is not EcmaAssembly ecmaAssembly)
+                        return Enumerable.Empty<CustomAttributeValue<TypeDesc>>();
+                    return ecmaAssembly.GetDecodedCustomAttributesForModule(attributeNamespace, attributeName)
+                        .Concat(ecmaAssembly.GetDecodedCustomAttributes(attributeNamespace, attributeName));
```

Comment:
Combining assembly and module attributes is _very subtle_.

I'd prefer this whole helper to be more local to logging - e.g. stash it in a ILCompiler.Logging.Extensions namespace that only logging implementation will have a `using` for.

We try to be more deliberate on what is accepted by the compiler in general.

Also we try to avoid allocating. Putting the allocating APIs into the global `ILCompiler` namespace will guide people towards the wrong APIs. It's fine to use APIs like this when generating warnings. There won't be millions of those. But the rest of the compiler should keep using the less convenient APIs.

Reply:
I was struggling with this... so thanks for some guidance. I will keep the property/event/module methods in the general extensions class (as they simply supplement the ones available from the type system already). And I will move the "get attributes from anything" to the warning suppression class alone. I'll get rid of the module/assembly handling completely and to it in the one place where it's actually used.